### PR TITLE
Handle out of memory error in Vec::write

### PR DESCRIPTION
**Pull request by LenaWil2**
The Vec::write trait has the possibility to return errors, and Vec has the possibility to check for allocation errors through the unstable [try_reserve](https://github.com/rust-lang/rust/issues/48043) method. I think that it would be possible for the methods in the write trait to return the out of memory errors to make it possible for the user to handle them. 

I don't know if the change is big enough to open an issue for, if it's needed, please say so. I'm still new to the Rust STD.

Potential issues:

- [ ] If unwrap is called on the write, it may panic instead of abort, which is the intended behaviour.
  * Unless, of course, [`oom=panic`](https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md#oompanic).
- [ ] [try_reserve](https://github.com/rust-lang/rust/issues/48043) hasn't stabilized yet.
- [ ] I didn't know which error to return, since the `Simple` error probably doesn't give enough information, and the `Custom` allocates memory.

Solutions for the last problem:
  1. Return a `Simple` error.
     * Problem: probably doesn't give enough information, since errors sometimes get passed around a lot. 
  2. Return a `Custom` error.
     * Problem: Allocates memory, which, if it fails, aborts, and makes change useless.
  3. Try to allocate and return a `Custom` error, and if it fails, return an `Simple` error.
     * Problem: Non-deterministic, which you don't want to be as a standard library.
  4. Same as 3, but always try to keep at least one pair of the two boxes needed to allocate one `Custom` error in thread local storage.
     * While slightly more deterministic than 3, it's ugly, still not 100% deterministic, and has a big overhead. 
  5. Same as 3, but add an `OOM` error to errorkind, so the user can check for this case. 
     * Calling `source` on the error is still non-deterministic, but the user has an consistent way to check for errors.

Currently this pull request uses solution 1.

### CODE REVIEW

File: src/libstd/io/impls.rs [hunk 1]
```diff
@@ -337,14 +339,21 @@ impl Write for &mut [u8] {
 impl Write for Vec<u8> {
     #[inline]
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if unlikely(self.try_reserve(buf.len()).is_err()) {
```
Comment (LenaWil) [inline src/libstd/io/impls.rs]: I'm not sure if I'm allowed to use unlikely

File: src/libstd/io/impls.rs [hunk 2]
```diff
@@ -337,14 +339,21 @@ impl Write for &mut [u8] {
...
 
     #[inline]
     fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
         let len = bufs.iter().map(|b| b.len()).sum();
-        self.reserve(len);
+
+        if unlikely(self.try_reserve(len).is_err()) {
+            return default_write_vectored(|buf| write_vec_oom(self, buf), bufs);
+        }
+
```
Comment (LenaWil) [inline src/libstd/io/impls.rs]: Minor style issue, but should I remove the empty lines?

File: src/libstd/io/impls.rs [hunk 3]
```diff
@@ -363,6 +375,26 @@ impl Write for Vec<u8> {
     }
 }
 
+fn write_vec_oom(vec: &mut Vec<u8>, buf: &[u8]) -> io::Result<usize> {
```
Comment (LenaWil) [inline src/libstd/io/impls.rs]: Also a minor style question: should I call this function `write_vec_oom` or `vec_write_oom`?

File: src/libstd/io/impls.rs [hunk 4]
```diff
@@ -363,6 +375,26 @@ impl Write for Vec<u8> {
...
+    debug_assert!(
+        !buf.is_empty(),
+        "write_vec_oom called with a empty buffer, should be handled earlier"
```
Comment (LenaWil) [inline src/libstd/io/impls.rs]: I was a bit in doubt if the message is really needed.

File: src/libstd/io/impls.rs [hunk 5]
```diff
@@ -363,6 +375,26 @@ impl Write for Vec<u8> {
...
+    );
+    // Using try_reserve instead of try_reserve_exact, which fails earlier, but the allocation time is O(1) amortized.
+    // try_reserve is guaranteed not to reallocate if there is enough space in the vector,
+    // and we need to write at least one byte to return successfully.
+    match vec.try_reserve(1) {
+        Ok(()) => {
+            let available_cap = vec.capacity() - vec.len();
+            debug_assert!(available_cap > 0);
+            let new_len = available_cap.min(buf.len());
```
Comment (LenaWil) [inline src/libstd/io/impls.rs]: Minor style question: `let new_len` or `let len`?

File: src/libstd/io/impls.rs [hunk 6]
```diff
@@ -337,14 +339,21 @@ impl Write for &mut [u8] {
...
+        if unlikely(self.try_reserve(buf.len()).is_err()) {
+            return write_vec_oom(self, buf);
+        }
         self.extend_from_slice(buf);
         Ok(buf.len())
     }
 
     #[inline]
     fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
         let len = bufs.iter().map(|b| b.len()).sum();
```
Comment (LenaWil) [inline src/libstd/io/impls.rs]: The code here assumes that `.sum()` won't overflow usize. I'm not sure this is an safe assumption, but `BufWriter` didn't handle it properly either, so I assumed it's fine.

**Comment by the8472** [comment]
>  I didn't know which error to return, since the Simple error probably doesn't give enough information, and the Custom allocates memory.

Another option would be `Error.from_raw_os_error(libc::ENOMEM)` since that's what you would be getting from mmap when you're OOM.

**Comment by LenaWil** [comment]
> Another option would be `Error.from_raw_os_error(libc::ENOMEM)` since that's what you would be getting from mmap when you're OOM.

Does that work on Windows?

**Comment by the8472** [comment]
Ah right. The constant exists https://docs.rs/libc/0.2.67/i686-pc-windows-msvc/libc/constant.ENOMEM.html
But I think it wouldn't be semantically correct since windows uses virtualalloc for allocations which doesn't try to be posix-compatible.

**Comment by retep998** [comment]
On Windows there are two similar error codes: `ERROR_NOT_ENOUGH_MEMORY` and `ERROR_OUTOFMEMORY`. I'm not entirely sure which is the right one to use.

**Comment by LenaWil** [comment]
I'm sorry for being so impatient, but @KodrAus, are you going to review this?

File: src/libstd/io/impls.rs [hunk 7]
```diff
@@ -353,6 +362,9 @@ impl Write for Vec<u8> {
 
     #[inline]
     fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
```
Comment (LenaWil) [inline src/libstd/io/impls.rs]: Now I'm thinking about this, maybe we should still write as much as possible, and thus fall back to the default `write_all` method. It's more consistent with the implementation of Write for slices.

**Comment by Amanieu** [comment]
I don't think this is a direction we want to go in for several reasons:
- This change adds quite a bit of complexity for a case that is in practice extremely rare.
- This is technically a behavior change on a stable interface: code which previously didn't need to deal with OOM (because it panicked/aborted) now needs to check the return value of `write`. If they don't check the result (which is reasonable since it previously never failed), then they will continue execution with an empty vector instead of what they expected it to contain.
- There is no precedent for this in the standard library, we typically don't return errors for OOM unless the user has explicitly asked for it.

cc @rust-lang/libs

**Comment by LukasKalbertodt** [comment]
I agree with @Amanieu. 

I know that there is a lot of code out there using this or similar APIs and assuming writing to the vector won't fail. In cases where the warning is just silenced with `let _ = ...` this code suddenly becomes incorrect. While I couldn't find documentation where we guarantee this behavior, I think the general assumption is that `std` will abort or panic on OOM. I know that I write almost all of my code with that assumption. Changing this is not a good idea IMO.

**Comment by crlf0710** [comment]
@Amanieu @LukasKalbertodt  Ping from triage. How about starting a fcp close or simply close this, to move this PR forward?

**Pull request closed without merging**